vllm/core/block_manager_v1.py

```diff
@@ -255,12 +255,26 @@ def __init__(
                 Device.CPU, block_size, num_cpu_blocks)
         # Mapping: seq_id -> BlockTable.
         self.block_tables: Dict[int, BlockTable] = {}
+        # Mapping: req_id -> BlockTable
+        # Note that each SequenceGroup has a unique
+        # request ID
+        self.cross_block_tables: Dict[str, BlockTable] = {}
+
+    def get_seq_num_required_blocks(self, seq: Sequence) -> int:
+        if seq is None:
+            return 0
+        return len(seq.logical_token_blocks)
 
     def can_allocate(self, seq_group: SequenceGroup) -> AllocStatus:
         # FIXME(woosuk): Here we assume that all sequences in the group share
         # the same prompt. This may not be true for preempted sequences.
-        seq = seq_group.get_seqs(status=SequenceStatus.WAITING)[0]
-        num_required_blocks = len(seq.logical_token_blocks)
+
+        self_num_required_blocks = self.get_seq_num_required_blocks(
+            seq_group.get_seqs(status=SequenceStatus.WAITING)[0])
+        cross_num_required_blocks = self.get_seq_num_required_blocks(
+            seq_group.get_encoder_seq())
+        num_required_blocks = self_num_required_blocks + \
+                              cross_num_required_blocks
 
         if self.block_sliding_window is not None:
             num_required_blocks = min(num_required_blocks,
```

Comment:
Would this code on line 279 still be valid for an encoder-decoder model with SWA?
```
        if self.block_sliding_window is not None:
            num_required_blocks = min(num_required_blocks,
                                      self.block_sliding_window)
```
Wouldn't you need twice as many blocks in this case?

Reply:
Thank you for your feedback! I would not focus on that too much, as at this time we are not committing to supporting encoder/decoder+SWA in this workstream (of course the existing SWA support for decoder-only models will be unchanaged)

Reply:
Perhaps it would be a good idea to raise some kind of "not implemented" error here then.

Reply:
Okay. Just pushed a commit to address this.